#160314849 Users should be able to categorise their articles

emmaadesile opened the pull request:
### What does this PR do?
Users can categorize their articles

### Description of tasks to be completed
- Admins can do CRUD operations on categories
- Users can get all categories
- Users can view all the articles in a category
- Authors can add their articles to a category
- Authors can remove their articles from any category

### How to manually test it
- After cloning the repo
- Install and configure Postgresql
- use .envsample to create your environment variables
- run `npm install` to install dependencies
- run `npm run migrate` to create the create tables in the db
- run `npm start:dev` to start the development server
- Admin can create `POST` a new `category/api/article-categories`
- Admin can update `PUT` an existing category `/api/article-categories/:categoryName`
- Admin can delete `DELETE` an existing category `/api/article-categories/:categoryName`
- A user or an author can view all categories `GET` `/api/article-categories`
- A user or an author can view all articles in a category GET /api/article-categories
- Authors can add their article(s) to a category using the articleId `POST` `/api/article-categories/:categoryName`

### What are the relevant pivotal tracker stories?
[160314849](https://www.pivotaltracker.com/story/show/160314849)

emmaadesile committed:
ft(categorize-articles): create models for categorising articles

- create category model
- add associations between article and category model
- create an articleCategories join table for article and category
- migrate the tables
- write tests and watch them fail

[Delivers 160314849]

emmaadesile committed:
ft(categorise-articles): create CRUD operations categories

- user can view all categories
- only admin can create a new category
- only admin can update a category
- only admin can delete a category

[Delivers 160314849]

emmaadesile committed:
ft(categorize-articles): write tests for endpoints for categorizing articles

[Delivers 160314849]

emmaadesile committed:
ft(categorize-articles): write tests for article-categories endpoints

- users cannot add an article to more than 3 categories

[Delivers 160314849]

emmaadesile committed:
ft(categorise-articles): write tests for the category model

- write tests for category model and the association

d-beloved reviewed (approved):
LGTM

CEOehis commented:
LGTM

CEOehis reviewed (approved):
LGTM

aknwosu commented on server/controllers/categoryController.js:
```diff
@@ -0,0 +1,366 @@
+   * @returns {object} The body of the response message
+   * @memberof CategoryController
+   */
+  static createCategory(req, res) {
```
this should only be accessible by admins

aknwosu commented on server/controllers/categoryController.js:
```diff
@@ -0,0 +1,366 @@
+    }
+    return Category.all()
+      .then((allCategories) => {
+        if (allCategories.length === 10) {
```
why is there a limit on how many categories can be added?

aknwosu commented on server/controllers/categoryController.js:
```diff
@@ -0,0 +1,366 @@
+          });
+        }
+        return Category.findOne({
+          where: { name: newCategory }
```
you probably want to sanitize this and make sure categories cannot be duplicated depending on the capitalization

aknwosu commented on server/controllers/categoryController.js:
```diff
@@ -0,0 +1,366 @@
+        }
+        // If category exists, check if article exists
+        Article.findOne(({
+          where: { title: articleTitle }
```
use article ID

aknwosu commented on server/controllers/categoryController.js:
```diff
@@ -0,0 +1,366 @@
+        if (!category) {
+          return res.status(404).json({
+            status: 'error',
+            error: 'Category does not exist'
```
error: 'There are no articles in this category'

aknwosu commented on server/controllers/categoryController.js:
```diff
@@ -0,0 +1,366 @@
+        }
+        // If category exists, check if article exists
+        Article.findOne(({
+          where: { title: articleTitle }
```
use articleID

aknwosu commented on server/controllers/categoryController.js:
```diff
@@ -0,0 +1,366 @@
+            if (article.authorId !== userId) {
+              return res.status(403).json({
+                status: 'error',
+                error: 'You cannot remove another author\'s article from this category',
```
"You do not have sufficient permissions for this action"

emmaadesile committed:
ft(categorise-articles): add ar

- add articles to category using article id
- remove articles using category using article id
- remove limit on the number of categories an admin create
- Fix failing tests

[Delivers 160314849]

emmaadesile committed:
ft(fix): remove ununsed utility function

emmaadesile committed:
ft(fix): make sure all tests are passing

emmaadesile committed:
ft(fix): fix errors in getAllCategoriesController

emmaadesile committed:
add more tests
